src/main/man/man1/mommy.1
.TH mommy "1" "%%VERSION_DATE%%" "mommy %%VERSION_NUMBER%%" "user commands"


.SH name
mommy - here to support you\(ti


.SH synopsis
.TP
\fBmommy\fP [\fIoptions\fP] \fIcommand\fP ...
e.g. \fBmommy\fP npm test
.TP
\fBmommy\fP [\fIoptions\fP] \fB-e\fP \fIcommand\fP
e.g. \fBmommy\fP -e "ls -l | wc -l"
.TP
\fBmommy\fP [\fIoptions\fP] \fB-s\fP \fIstatus\fP
e.g. \fBmommy\fP -s $?


.SH description
.UR https://github.com/fwdekker/mommy/tree/v%%VERSION_NUMBER%%#readme
.UE


.SH options
the following options are just for invoking \fBmommy\fP.
to change \fBmommy\fP's output, you should change the config file.
for more info about config files, check
.UR https://github.com/fwdekker/mommy/tree/v%%VERSION_NUMBER%%#configuration
.UE .

.SS run
.TP
\fIcommand\fP
write your \fIcommand\fP after \fBmommy\fP, and \fBmommy\fP will run and respond to the \fIcommand\fP based on its exit code\(ti
.TP
\fB-e\fP \fIcommand\fP, \fB--eval=\fP\fIcommand\fP
runs \fIcommand\fP and responds to the output.
note that \fIcommand\fP should be given as a single argument\(ti
.TP
\fB-p\fP, \fB--pipefail\fP
when using \fB-e\fP or \fB--eval\fP, this option enables pipefail in the
evaluated command.
works only if \fB/bin/sh\fP on your system supports pipefail\(ti
.TP
\fB-s\fP \fIstatus\fP, \fB--status=\fP\fIstatus\fP
gives output corresponding to exit code \fIstatus\fP\(ti

.SS output control
.TP
\fB-1\fP
writes output to \fBstdout\fP instead of \fBstderr\fP\(ti
.TP
\fB-t\fP, \fB--toggle\fP
toggles whether \fBmommy\fP should display output at all.
applies to all sessions of all shells of the current user, until this option is
toggled again.
useful if you want to temporarily silence \fBmommy\fP without editing your
config files\(ti

.SS configuration
.TP
\fB-d\fP \fIdirs\fP, \fB--global-config-dirs=\fP\fIdirs\fP
sets global configuration dirs to the colon-separated list in \fIdirs\fP.
\fBmommy\fP will traverse this list in order and load the first config file 'mommy/config.sh' that exists.
if not set, \fBmommy\fP will use '$XDG_CONFIG_DIRS:/etc/:/usr/local/etc/' as the default value of \fIdirs\fP\(ti
.TP
\fB-u\fP \fIdir\fP, \fB--user-config-dir=\fP\fIdir\fP
if set, \fBmommy\fP will load config file '\fIdir\fP/mommy/config.sh' if it exists.
if not set, \fBmommy\fP will use '$XDG_CONFIG_HOME' or '$HOME/.config' as the default value of \fIdir\fP\(ti
.TP
\fB-r\fP \fIname\fP, \fB--role=\fP\fIname\fP
if set, \fBmommy\fP will load 'mommy/roles/\fIname\fP.sh' instead of 'mommy/config.sh' in the various config directories\(ti

.SS about
.TP
\fB-h\fP, \fB--help\fP
opens \fBmommy\fP's manual page\(ti
.TP
\fB-v\fP, \fB--version\fP
displays \fBmommy\fP's version information\(ti


.SH bugs
.UR https://github.com/fwdekker/mommy/issues
.UE


CHANGELOG.md
# Changelog
## [9.9.9-unreleased] -- 9999-12-31
### breaking
* 🔧 mommy no longer supports `-c`/`--config` arguments. instead, to specify a specific config file, use a combination of `-u`/`--user-config-dir` and `-d`/`--global-config-dirs` to specify the directory to search in, and `-r`/`--role` to specify the file. see the readme for full details~ (#TODO)

### added
* 🎭 mommy now supports roles (aka moods/profiles) so you can easily swap between several config files. to add a role `foo`, place your role config in `$HOME/.config/mommy/roles/foo.sh`, and run mommy with `-r foo` or `--role=foo`~
* 💵 mommy now includes bash completions~ ([#43](https://github.com/fwdekker/mommy/issues/43)) ([#170](https://github.com/fwdekker/mommy/issues/170))
* 🫂 mommy now supports configuring `MOMMY_SWEETIE_ENCOURAGE` to override the value of `MOMMY_SWEETIE` only in encouragements~ ([#183](https://github.com/fwdekker/mommy/pull/183))
* #️⃣ mommy now supports templates starting with `#` using `%%#%%`~ ([#173](https://github.com/fwdekker/mommy/issues/173)) ([#174](https://github.com/fwdekker/mommy/issues/174))
* 🌞 mommy once again distributes a package for solaris, attached as a `.p5p` file to each github release~ ([#176](https://github.com/fwdekker/mommy/issues/176))
* 🍱 mommy now also runs tests on busybox~ (#TODO)


## [1.8.0] -- 2025-12-03
### added
* 🪈 mommy now enables [pipefail](https://stackoverflow.com/q/68465355/) for you when you use `-e`/`--eval` if you also pass the `-p`/`--pipefail` option. if your shell at `/bin/sh` does not support pipefail, you'll receive an error if you do this~ ([#160](https://github.com/fwdekker/mommy/pull/160)) ([#165](https://github.com/fwdekker/mommy/pull/165))
* ✍️ mommy added a package for the haiku operating system~ ([#167](https://github.com/fwdekker/mommy/pull/167))

### changed
* 🦈 mommy updated fish completions for fish v4~ ([#166](https://github.com/fwdekker/mommy/pull/166))
* 📒 mommy improved her manual page~ ([#166](https://github.com/fwdekker/mommy/pull/166))

### fixed
* 🌿 mommy fixed builds on opensuse tumbleweed~ ([#162](https://github.com/fwdekker/mommy/pull/162))
* 🚧 mommy now detects when the new toggle feature (`-t`/`--toggle`) from v1.7.0 fails to (de)activate~ ([#161](https://github.com/fwdekker/mommy/pull/161)) ([#163](https://github.com/fwdekker/mommy/pull/163))
* 👩‍🔬 mommy now keeps config files in unit tests completely separate from real config files~ ([#162](https://github.com/fwdekker/mommy/pull/162))


## [1.7.0] -- 2025-12-01
### added
* 🫥 mommy can now be toggled on or off without changing any configuration files using the `-t` or `--toggle` option. just run `mommy -t` and you'll no longer get any output from mommy, until you run `mommy -t` again. the toggle setting is per user, but carries across shells and even reboots. mommy thanks [amber sprenkels](https://github.com/dsprenkels) for suggesting the feature, and thanks [david scann](https://github.com/fwdekker/mommy/issues/46#issuecomment-2888498320) for additional feedback~ ([#46](https://github.com/fwdekker/mommy/pull/46))

### changed
* ✍️ mommy changed the capitalisation of its maintainer's username~ ([#152](https://github.com/fwdekker/mommy/pull/152))

### security
* 🔧 mommy improved the security of her automated test scripts~ ([#152](https://github.com/fwdekker/mommy/pull/152))


## [1.6.0] -- 2025-03-09
### added
* 📇 mommy now supports independent possessive (`theirs`) and reflexive (`themself`) pronouns. to use these new forms inside a compliment or encouragement, write `%%THEIRS%%` or `%%THEMSELF%%`, respectively. to choose mommy's pronouns, you should now write `MOMMY_PRONOUNS="they them their theirs theirself"` in your config instead of `MOMMY_PRONOUNS="they them their"`. specifying all five pronoun forms is recommended, but only specifying three forms still works and is _not_ deprecated. for more details, see [the list of all settings](https://github.com/fwdekker/mommy/tree/v1.6.0?tab=readme-ov-file#list-of-all-settings) and [how to configure templates](https://github.com/fwdekker/mommy/tree/v1.6.0?tab=readme-ov-file#how-to-configure-templates)~
* ☄️ mommy now supports using whitespace inside pronouns using `%%_%%`~
* 🚀 mommy now explains how to configure [starship](https://starship.rs/)~ ([#135](https://github.com/fwdekker/mommy/pull/135))
* ☕ mommy explains how to use fish completions on macos~

### fixed
* 👹 mommy ensures freebsd builds don't have the target version hardcoded~


## [1.5.1] -- 2024-09-28
### added
* 🤸‍♀️ mommy now supports `-d` as an alias of `--global-config-dirs`~

### changed
* 👤 mommy now refers to the user by their username by default~ ([#131](https://github.com/fwdekker/mommy/issues/131))


## [1.5.0] -- 2024-02-28
### added
* 🦓 mommy now supports templates with literal slashes using `%%S%%`~ ([#107](https://github.com/fwdekker/mommy/issues/107))
* 🐟 mommy added completions for fish and zsh for the new options introduced in v1.4.0~ ([#105](https://github.com/fwdekker/mommy/issues/105))

### changed
* 🙅‍♀️ mommy's `MOMMY_FORBIDDEN_WORDS` setting now interprets each word as a regex~ ([#103](https://github.com/fwdekker/mommy/issues/103))
* 🚀 mommy is now 4 times as fast~ ([#106](https://github.com/fwdekker/mommy/issues/106))


## [1.4.0] -- 2024-02-17
### added
* 🌍 mommy now supports a global config file that applies to all users, stored for example in `/usr/mommy/config.sh`~ ([#95](https://github.com/fwdekker/mommy/issues/95)) ([#96](https://github.com/fwdekker/mommy/issues/96))
* 📏 mommy accepts long command-line options for all options (like `--config=<file>` for `-c <file>`)~

### changed
* 🕳️ mommy now interprets an empty config file path (`-c`) as you not wanting to use a config file~
* ✅ mommy improved her input validation, giving clearer error messages when something is wrong~
* 🏓 mommy improved the links to and from the table of contents in her readme~

src/main/completions/bash/mommy.bash
## Four basics to writing Bash completions:
# COMPREPLY - array for possible completions
# COMP_WORDS - array of individual arguments typed so far
# COMP_CWORD - index of the command argument at cursor
# COMP_LINE - current command line

_mommy()
{
	local args_avail=( \
	    "-h" "--help" \
	    "-t" "--toggle" \
	    "-v" "--version" \
	    "-e" "--eval=" \
	    "-p" "--pipefail" \
	    "-s" "--status=" \
	    "-d" "--global-config-dirs=" \
	    "-u" "--user-config-dir=" \
	    "-r" "--role=" \
    )
    local cur="${COMP_WORDS[COMP_CWORD]}"

    if [[ "$COMP_CWORD" -eq 1 ]]; then
        mapfile -t COMPREPLY < <(compgen -W "${args_avail[*]}" -- "$cur")
        return
    fi

    flag=0

    while [[ "$flag" -eq 0 ]]; do
        local argument="${COMP_WORDS[ ((COMP_CWORD - 1)) ]}"
        if [[ "$COMP_CWORD" -ge 2 ]]; then
            case "$argument" in
                "-h" | "--help")
                    args_avail=("${args_avail[@]/-h}")
                    args_avail=("${args_avail[@]/--help}")
                    flag=1 ;;
                "-d" | "--global-config-dirs=")
                	args_avail=("${args_avail[@]/-d}")
                    args_avail=("${args_avail[@]/--global-config-dirs=}")
                    mapfile -t COMPREPLY < <(compgen -A directory -- "$cur") ;;
                "-r" | "--role=")
                	args_avail=("${args_avail[@]/-r}")
                    args_avail=("${args_avail[@]/--role=}")
                    COMPREPLY=() ;;
                "-e" | "--eval=")
                	args_avail=("${args_avail[@]/-e}")
                	args_avail=("${args_avail[@]/--eval=}")
                    mapfile -t COMPREPLY < <(compgen -A command -- "$cur") ;;
                "-s" | "--status=")
               		args_avail=("${args_avail[@]/-s}")
               		args_avail=("${args_avail[@]/--status=}")
                    COMPREPLY=("0" "1") ;;
                "-t" | "--toggle")
                    args_avail=("${args_avail[@]/-t}")
                    args_avail=("${args_avail[@]/--toggle}")
                    flag=1 ;;
                "-p" | "--pipefail")
               		args_avail=("${args_avail[@]/-p}")
                	args_avail=("${args_avail[@]/--pipefail}")
                    mapfile -t COMPREPLY < <(compgen -W "${args_avail[*]}" -- "$cur") ;;
            esac
            return
        fi
    done
}
complete -F _mommy mommy
